### PR TITLE
[SPARK-25768][SQL] fix constant argument expecting UDAFs

### PR DESCRIPTION
**Pull request by peter-toth**
## What changes were proposed in this pull request?

Without this PR some UDAFs like `GenericUDAFPercentileApprox` can throw an exception because expecting a constant parameter (object inspector) as a particular argument.

The exception is thrown because `toPrettySQL` call in `ResolveAliases` analyzer rule transforms a `Literal` parameter to a `PrettyAttribute` which is then transformed to an `ObjectInspector` instead of a `ConstantObjectInspector`.
The exception comes from `getEvaluator` method of `GenericUDAFPercentileApprox` that actually shouldn't be called during `toPrettySQL` transformation. The reason why it is called are the non lazy fields in `HiveUDAFFunction`.

This PR makes all fields of `HiveUDAFFunction` lazy.

## How was this patch tested?

added new UT

### CODE REVIEW

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveUDFSuite.scala [hunk 1]
```diff
@@ -638,6 +638,21 @@ class HiveUDFSuite extends QueryTest with TestHiveSingleton with SQLTestUtils {
         Row(3) :: Row(3) :: Nil)
     }
   }
+
+  test("constant argument expecting Hive UDF") {
```
Comment (mgaido91) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveUDFSuite.scala]: may you please reference the JIRA?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveUDFSuite.scala [hunk 2]
```diff
@@ -638,6 +638,21 @@ class HiveUDFSuite extends QueryTest with TestHiveSingleton with SQLTestUtils {
...
+    val testData = spark.range(10).toDF()
+    withTempView("inputTable") {
+      testData.createOrReplaceTempView("inputTable")
```
Comment (mgaido91) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveUDFSuite.scala]: nit: we can move here `spark.range(10)`

**Review by mgaido91 — COMMENTED** [review]
@peter-toth may you please elaborate a bit more the change in the PR description? What happened before the PR? Why? Thanks.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/hiveUDFs.scala [hunk 3]
```diff
@@ -339,40 +339,38 @@ private[hive] case class HiveUDAFFunction(
     val parameterInfo = new SimpleGenericUDAFParameterInfo(inputInspectors, false, false)
     resolver.getEvaluator(parameterInfo)
   }
+  
+  private case class Mode(evaluator: GenericUDAFEvaluator, objectInspector: ObjectInspector)
```
Comment (peter-toth) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/hiveUDFs.scala]: evaluator and object inspector need to be initialized together

Comment (mgaido91) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/hiveUDFs.scala]: nit: maybe we can find a better name for this class

**Review by mgaido91 — COMMENTED** [review]
the change makes sense to me. I have just one minor comment about naming. 

cc @cloud-fan @dongjoon-hyun do you think we can trigger a build?

**Comment by peter-toth** [comment]
cc @cloud-fan I believe this is a regression because https://issues.apache.org/jira/browse/SPARK-18186

**Comment by dongjoon-hyun** [comment]
ok to test

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by mgaido91** [comment]
retest this please

**Comment by mgaido91** [comment]
LGTM

**Comment by cloud-fan** [comment]
thanks, merging to master/2.4/2.3!

**Pull request closed without merging**